### ISSUE TITLE
Problems building rviz from source [foxy]

### ISSUE THREAD

**Issue by mnissov** [ros2/rviz#784]
With recent problems concerning rviz and showing pointclouds, [782](https://github.com/ros2/rviz/issues/782), I am attempting to build the updated version of rviz while waiting for an official release.

I build rviz by according to the `README.md` by
```bash
mkdir -p ~/rviz2_ws/src
cd ~/rviz2_ws/src
git clone -b foxy https://github.com/ros2/rviz.git
cd ../
colcon build --merge-install
```
and most packages are able to build, except for `rviz_default_plugins`. The error code is the following
```bash
$ colcon build --merge-install
Starting >>> rviz_assimp_vendor
Starting >>> rviz_ogre_vendor
Finished <<< rviz_assimp_vendor [0.14s]                                                                
Finished <<< rviz_ogre_vendor [0.21s]                  
Starting >>> rviz_rendering
Finished <<< rviz_rendering [0.28s]                     
Starting >>> rviz_common
Starting >>> rviz_rendering_tests
Finished <<< rviz_rendering_tests [0.18s]                                                               
Finished <<< rviz_common [0.74s]                     
Starting >>> rviz_visual_testing_framework
Finished <<< rviz_visual_testing_framework [0.19s]                  
Starting >>> rviz_default_plugins
[Processing: rviz_default_plugins]                             
--- stderr: rviz_default_plugins                               
/usr/bin/ld: CMakeFiles/point_cloud2_display_test.dir/test/rviz_default_plugins/displays/pointcloud/point_cloud2_display_test.cpp.o: in function `rviz_common::MessageFilterDisplay<sensor_msgs::msg::PointCloud2_<std::allocator<void> > >::incomingMessage(std::shared_ptr<sensor_msgs::msg::PointCloud2_<std::allocator<void> > const>)':
point_cloud2_display_test.cpp:(.text._ZN11rviz_common20MessageFilterDisplayIN11sensor_msgs3msg12PointCloud2_ISaIvEEEE15incomingMessageESt10shared_ptrIKS5_E[_ZN11rviz_common20MessageFilterDisplayIN11sensor_msgs3msg12PointCloud2_ISaIvEEEE15incomingMessageESt10shared_ptrIKS5_E]+0x59): undefined reference to `rviz_common::_RosTopicDisplay::typeErasedMessageTaken(std::shared_ptr<void const>)'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/point_cloud2_display_test.dir/build.make:240: point_cloud2_display_test] Error 1
make[1]: *** [CMakeFiles/Makefile2:384: CMakeFiles/point_cloud2_display_test.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
/usr/bin/ld: CMakeFiles/point_cloud_scalar_display_test.dir/test/rviz_default_plugins/displays/pointcloud/point_cloud_scalar_display_test.cpp.o: in function `rviz_common::MessageFilterDisplay<sensor_msgs::msg::Illuminance_<std::allocator<void> > >::incomingMessage(std::shared_ptr<sensor_msgs::msg::Illuminance_<std::allocator<void> > const>)':
point_cloud_scalar_display_test.cpp:(.text._ZN11rviz_common20MessageFilterDisplayIN11sensor_msgs3msg12Illuminance_ISaIvEEEE15incomingMessageESt10shared_ptrIKS5_E[_ZN11rviz_common20MessageFilterDisplayIN11sensor_msgs3msg12Illuminance_ISaIvEEEE15incomingMessageESt10shared_ptrIKS5_E]+0x59): undefined reference to `rviz_common::_RosTopicDisplay::typeErasedMessageTaken(std::shared_ptr<void const>)'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/point_cloud_scalar_display_test.dir/build.make:225: point_cloud_scalar_display_test] Error 1
make[1]: *** [CMakeFiles/Makefile2:595: CMakeFiles/point_cloud_scalar_display_test.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
---
Failed   <<< rviz_default_plugins [41.3s, exited with code 2]

Summary: 6 packages finished [42.9s]
  1 package failed: rviz_default_plugins
  1 package had stderr output: rviz_default_plugins
  1 package not processed
```
**EDIT:** Using ubuntu 20.04 LTS and ros2 foxy

**Comment by jacobperron** [comment]
I'm pretty sure this build error is in part due to the revert done here, which removes API: https://github.com/ros2/rviz/pull/783

I would have still expected the build to work, however I think there is an issue with the linker ordering. It seems that the version of `rviz_default_plugins` in the workspace is trying to link against the version on the system. Can you try removing the installed version of rviz and confirm that you can build?

        sudo apt remove ros-foxy-rviz*

**Comment by jacobperron** [comment]
I don't know if it will fix this issue, but it's probably worth backporting https://github.com/ros2/rviz/pull/727 to Foxy as it resolves a similar linking issue.

**Comment by Enri2077** [comment]
It works!
Initially I tried to build rviz2 from source to work around https://github.com/ros2/rviz/issues/782 and got the same build error as @mnissov.
I uninstalled all rviz packages `sudo apt remove ros-foxy-rviz*`, cloned the foxy branch in my workspace and now it builds and works as expected.
I can visualize maps and laser scans. I don't have any pointclouds at hand, but I assume those would be working as well.
